.github/workflows/main.yml

```
@@ -58,10 +69,16 @@ jobs:
           . venv/bin/activate
           PYTHONHASHSEED=0 ./scripts/gen_verilog.py --verbose --config full
 
-      - uses: actions/download-artifact@v3
+      - uses: actions/cache@v3
+        env:
+          cache-name: cache-regression-tests
         with:
-          name: "riscv-tests"
-          path: test/external/riscv-tests
+          submodules: recursive
+          path: test/external/riscv-tests/test-*
+          key: ${{ env.cache-name }}-${{ runner.os }}-${{
+            needs.build-regression-tests.outputs.container_id }}-${{ hashFiles('**/test/external/riscv-tests') }}
+          restore-keys: |
+            ${{ env.cache-name }}-${{ runner.os }}-${{ needs.build-regression-tests.outputs.container_id }}-
```

```suggestion
          fail-on-cache-miss: true
```
I think that failing on cache miss would be helpful here.
"Run" action always runs after "Build" action, so the cache with built test files must be already present (and we want to use that exact cache).

This would also prevent reuploading the cache (and all weird problems with it).

In this case `restore-keys` would not be needed.